src/socket.rs

```diff
@@ -1803,26 +1852,58 @@ mod tests {
                     let socket = unwrap!(UdpSocket::bind(&addr!("127.0.0.1:0"), &handle));
                     let inner = Inner::new_shared(&handle, socket, listener_set_readiness);
 
-                    let (packet_rx, remote_peer_addr) = wait_for_packet(&handle);
+                    let (packets_rx, remote_peer_addr) = wait_for_packets(&handle);
+
                     let mut packet = Packet::reset();
                     packet.set_connection_id(12_345);
-
-                    unwrap!(unwrap!(inner.write()).process_unknown(
-                        packet,
-                        remote_peer_addr,
-                        &inner
-                    ));
-
-                    packet_rx
-                        .with_timeout(Duration::from_secs(1), &handle)
-                        .map(move |res_opt| {
-                            drop(inner);
-                            res_opt.is_none()
-                        })
+                    unwrap!(
+                        unwrap!(inner.write())
+                            .reset_packets
+                            .push_back((packet, remote_peer_addr))
+                    );
+                    let mut packet = Packet::reset();
+                    packet.set_connection_id(23_456);
+                    unwrap!(
+                        unwrap!(inner.write())
+                            .reset_packets
+                            .push_back((packet, remote_peer_addr))
+                    );
+
+                    // keep retrying until all packets are sent out
+                    future::poll_fn(move || unwrap!(inner.write()).flush_reset_packets())
+                        .and_then(move |_| packets_rx.take(2).collect().map_err(|e| panic!(e)))
                 });
+                let packets = unwrap!(evloop.run(task));
+
+                let packet_ids: Vec<u16> = packets.iter().map(|p| p.connection_id()).collect();
+                assert_that!(&packet_ids, contains(vec![12_345, 23_456]).exactly());
```

Comment:
I still don't really get what this hamcrest stuff is for. Why not just write

    assert_eq!(packet_ids, vec![12_345, 23_456])

Or, if order matters,

    assert_eq!(packet_ids.into_iter().collect(), hash_set!{12_345, 23_4564})

Reply:
Probably you meant "if order doesn't matter" in case of hashset example :)
In this case order does not matter to me and it's not predictable.
In Rust this fails:
```rust
    assert_eq!(vec![1, 2, 3], vec![1, 3, 2]);
```

So I could use hashset with some collections hustling. But it's all about readability - it's imperative versus declarative. hamcrest is a set of useful/common matchers that reduce boilerplate code and make code a bit more readable.

Reply:
Okay, I'm not sold on it being more readable but if you wanna use it I won't stand in your way.